src/Agent.Worker/DiagnosticLogManager.cs

```diff
@@ -0,0 +1,262 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Text;
+using System.Runtime.InteropServices;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Agent.Worker;
+using Microsoft.VisualStudio.Services.Agent.Worker.Build;
+using Microsoft.VisualStudio.Services.WebApi;
+using Build2 = Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.Win32;
+using System.Diagnostics;
+using System.Linq;
+using System.Collections.ObjectModel;
+using System.Globalization;
+
+namespace Microsoft.VisualStudio.Services.Agent
+{
+    [ServiceLocator(Default = typeof(DiagnosticLogManager))]
+    public interface IDiagnosticLogManager : IAgentService
+    {
+        void UploadDiagnosticLogs(IExecutionContext executionContext, 
+                                  AgentJobRequestMessage message, 
+                                  DateTime jobStartTimeUtc);
+    }
+
+    // This class manages gathering data for support logs, zipping the data, and uploading it.
+    // The files are created with the following folder structure:
+    // ..\_layout\_work\_temp
+    //      \[job name]-support (supportRootFolder)
+    //          \files (supportFolder)
+    //              ...
+    //          support.zip
+    public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
+    {
+        public void UploadDiagnosticLogs(IExecutionContext executionContext, 
+                                         AgentJobRequestMessage message, 
+                                         DateTime jobStartTimeUtc)
+        {
+            executionContext.Debug("Starting diagnostic file upload.");
+
+            // Setup folders
+            // \_layout\_work\_temp\[jobname-support]
+            executionContext.Debug("Setting up diagnostic log folders.");
+            string tempDirectory = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Work), Constants.Path.TempDirectory);
+            ArgUtil.NotNullOrEmpty(tempDirectory, nameof(tempDirectory));
```

Comment:
ArgUtil.Directory(tempDirectory, nameof(tempDirectory)); also did NotNullOrEmpty Check.

Reply:
Fixed. Not sure why it isn't showing here.